### PR TITLE
Spin up a kind cluster and configure metallb

### PR DESCRIPTION
**Pull request by sarahalsmiller**
### Changes proposed in this PR:

- Empty test runner added to testing/conformance that can be invoked with go test ./... -tags conformance
- Set up a conformance test stack intesting/conformanc that installs metallb in addition to the dependencies required for the e2e test (though potentially they don't need all the same dependencies installed? Feedback needed there.) 
- Added a github action config that invokes go test ./... -tags conformance 

### How I've tested this PR:
go test ./... -tags conformance

### How I expect reviewers to test this PR:
go test ./... -tags conformance

### Checklist:
- [ X ] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use imperative present tense (e.g. Add support for...)

### CODE REVIEW

File: internal/testing/conformance/run_conformance_test.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"os"
+	"testing"
+
+	"github.com/hashicorp/consul-api-gateway/internal/testing/e2e"
+	"sigs.k8s.io/e2e-framework/pkg/env"
+)
+
+//This file is intended to help faciliate an easy and local way of running the
```
Comment (nathancoleman) [inline internal/testing/conformance/run_conformance_test.go:14]:
```suggestion
//This file is intended to help facilitate an easy and local way of running the
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -113,3 +113,42 @@ jobs:
...
+        curl -L https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64 -o ./kind
+        chmod +x ./kind
+        mv ./kind /usr/local/bin/kind
+        curl -L https://dl.k8s.io/release/v1.22.0/bin/linux/amd64/kubectl -o ./kubectl
+        chmod +x ./kubectl
+        mv ./kubectl /usr/local/bin/kubectl
+        curl -L https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize%2Fv4.4.0/kustomize_v4.4.0_linux_amd64.tar.gz -o kustomize_v4.4.0_linux_amd64.tar.gz
+        tar xvzf kustomize_v4.4.0_linux_amd64.tar.gz
+        mv kustomize /usr/local/bin/kustomize
+        rm kustomize_v4.4.0_linux_amd64.tar.gz
```
Comment (nathancoleman) [inline .github/workflows/ci.yml:139]: This has me wondering if we could use something like https://github.com/helm/kind-action in the future, but it is Actions-specific so makes our test setup less portable

File: internal/testing/conformance/run_conformance_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+	testenv.Run(m)
+}
+
+func TestRunConformanceTestSuite(t *testing.T) {
+	//TODO wait for gateway to be ready
+
+	//TODO invoke conformance tests
+
+}
```
[the remark this answers is not in the document]
Comment (Ramvling) [reply internal/testing/conformance/run_conformance_test.go:48]: Hey Nathan, I've been playing around with running it that way, pulling down their repo and execing out to the go test. It's kinda working, but its erroring before the test actually runs. I can show you what I mean when we hop on a call, but yes I'd also be interested in how everyone else in the kube-sig chat are running these conformance tests.

[the remark this answers is not in the document]
Comment (sarahalsmiller) [reply internal/testing/conformance/run_conformance_test.go:48]: Hey Nathan, I've been playing around with running it that way, pulling down their repo and execing out to the go test. It's kinda working, but its erroring before the test actually runs. I can show you what I mean when we hop on a call, but yes I'd also be interested in how everyone else in the kube-sig chat are running these conformance tests.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sarahalsmiller) [reply .github/workflows/ci.yml:139]: If we did something like that, I suppose we could also have something to run the test locally? Using kind-action would at a minimum cut down on our install time?

Comment (nathancoleman) [reply .github/workflows/ci.yml:139]: Meant to specify that this was just me thinking out loud, not something that needs to be done in this PR

File: internal/testing/conformance/run_conformance_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewstucki) [reply internal/testing/conformance/run_conformance_test.go:48]: From the chatter in the k8s slack, seems like other implementors (i.e. folks like Istio) _are_ essentially just re-writing the upstream `conformance_test.go` and using that after adding in their custom setup logic. TBH wondering if that might be simpler in our case due to the need to patch the upstream manifests -- especially since the suite subpackage accepts an option to [override the base applied manifests](https://github.com/kubernetes-sigs/gateway-api/blob/704c253f29c176ee2e396a5643f4e121018d3b6c/conformance/utils/suite/suite.go#L63).

@nathancoleman @sarahalsmiller what do y'all think about doing something like just duping the upstream `conformance_test.go` and doing something like rendering out the patched manifests via `kustomize` to some temporary file location and then passing in that temporary file to the `BaseManifests` argument there?

[the remark this answers is not in the document]
Comment (sarahalsmiller) [reply internal/testing/conformance/run_conformance_test.go:48]: Hey Andrew, @nathancoleman and I were just talking through this. Essentially I think we like the idea of not having to care what the upstream conformance_test.go looks like, but I also don't want to introduce any additional over head if I don't have to. I'm pretty indifferent either way and I'm happy to implement whichever runner the team thinks is better.

File: internal/testing/conformance/run_conformance_test.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// +build conformance
+
+package conformance
+
+import (
+	"os"
+	"testing"
+
+	"github.com/hashicorp/consul-api-gateway/internal/testing/e2e"
+	"sigs.k8s.io/e2e-framework/pkg/env"
```
Comment (andrewstucki) [inline internal/testing/conformance/run_conformance_test.go:11]: So, I'm _somewhat_ indifferent on this front, but personally I'd probably prefer not to mix/match testing frameworks, since they're both rather opinionated in their setup. I'm wondering if maybe first pass we can just work on the assumption that we have a functional `kubernetes` cluster at this stage (with, in the case of `kind`, already has `MetalLB` up and running) and potentially even the gateway controller already installed on the system (which we could make easier down the line if we so desire).

That way this would only involve the _much_ smaller task of rendering out the patched manifests and invoking the underlying conformance suite code.

Much of the setup logic for CI could then just get moved to the yaml manifest and we could keep the `conformance_test.go` parts closer the vanilla upstream. If we wanted to refactor down the line to do more of the setup in `go` that'd be cool too, but I figure that might be able to wait a bit.

Comment (sarahalsmiller) [reply internal/testing/conformance/run_conformance_test.go:11]: If we were to do it that way and pull it all out into the CI work, we could just use the kind action Nathan linked and simplify a lot of things. I'm pretty indifferent too, I just thought we wanted it all to be contained in `go test`. Definitely happy to spend some time proving out this other way though.

Comment (andrewstucki) [reply internal/testing/conformance/run_conformance_test.go:11]: I'm ok without it being _fully_ `go test`able at the get-go, so long as the major patching bits an actual test invocation are done in code at the beginning and `go test`able.

I'm also ok extending some of this work down-the-line to do the same random env spin-up we do in our `e2e` tests while just doing the minimal thing here. I think it'd be fine to hold off on doing the whole go-based `kind` spin-up in this PR, especially as we've talked about changing the way some of our e2e environments get provisioned to be more extensible for cloud systems and not just purely `kind`-based.

File: internal/testing/conformance/run_conformance_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewstucki) [reply internal/testing/conformance/run_conformance_test.go:48]: So, from reading through some of the evolution of the conformance testing framework, it looks like the sig wants to make it simple for people to actually import the suite as a go module and test it via a simple re-implementation of `conformance_test.go`: https://github.com/kubernetes-sigs/gateway-api/pull/1023

[the remark this answers is not in the document]
Comment (nathancoleman) [reply internal/testing/conformance/run_conformance_test.go:48]: The structure of the code and documentation had implied to me that the order of preference was:
1. `make conformance`
2. `go test ./conformance`
3. import the suite, rewrite upstream's `conformance_test.go`

I think the _cleanest_ option is the third if upstream maintainers are okay with that. We'll just need to keep an eye out for shifts in the suite, and I'm not entirely sure they plan on having those be backwards compatible.

Just asked the question in kube-sig Slack [here](https://kubernetes.slack.com/archives/CR0H13KGA/p1648590160555709?thread_ts=1648570950.946279&cid=CR0H13KGA).

[the remark this answers is not in the document]
Comment (nathancoleman) [reply internal/testing/conformance/run_conformance_test.go:48]: Definitely feeling better about rewriting `conformance_test.go` now

image
